spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/MicrometerHolder.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.amqp.rabbit.listener;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import org.springframework.beans.factory.config.BeanDefinition;
+import org.springframework.beans.factory.support.BeanDefinitionRegistry;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.ConfigurableApplicationContext;
+import org.springframework.lang.Nullable;
+
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Timer;
+import io.micrometer.core.instrument.Timer.Builder;
+import io.micrometer.core.instrument.Timer.Sample;
+
+/**
+ * Abstraction to avoid hard reference to Micrometer.
+ *
+ * @author Gary Russell
+ * @since 2.4.6
+ *
+ */
+final class MicrometerHolder {
+
+	private final ConcurrentMap<String, Timer> timers = new ConcurrentHashMap<>();
+
+	private final MeterRegistry registry;
+
+	private final Map<String, String> tags;
+
+	private final String listenerId;
+
+	MicrometerHolder(@Nullable ApplicationContext context, String listenerId, Map<String, String> tags) {
+		if (context == null) {
+			throw new IllegalStateException("No micrometer registry present");
+		}
+		Map<String, MeterRegistry> registries = context.getBeansOfType(MeterRegistry.class, false, false);
+		registries = filterRegistries(registries, context);
```

Comment:
I wonder why does `ObjectProvider.getIfUnique()` not work for us here:
```

	/**
	 * Return an instance (possibly shared or independent) of the object
	 * managed by this factory.
	 * @return an instance of the bean, or {@code null} if not available or
	 * not unique (i.e. multiple candidates found with none marked as primary)
	 * @throws BeansException in case of creation errors
	 * @see #getObject()
	 */
	@Nullable
	T getIfUnique() throws BeansException;
```
?

Reply:
Where do we get one from?

Reply:
`ApplicationContext.getBeanProvider(Class<T> requiredType)`

Reply:
Oh, I didn't know about that; I've only seen them in the context of auto wiring. Cool.